[DCP] Cache save plans in default planner

saumishr opened the pull request:
Summary:
This PR caches the save plans to significantly reduce the collective cost for successive checkpoint save attempts. Here is the high level approach:
-  Create the local plan and cache the same.
- In next iteration, compare the local plan with the cached plan metadata. If no change, do not send that local plan in the collective.
- Global plan step, will only create the global plan with the new delta plans and empty plans for the cached ones. 
- Finish plan step will check for the empty plans. If its empty, it will grab the cached plan. If not, it will use the new plan provided.

Test Plan: UTs

Differential Revision: D69224491

## How to enable the caching:
DefaultSavePlanner introduces the enable_plan_caching which is set to False by default for now. 
https://github.com/pytorch/pytorch/pull/147343/files#diff-579bbb7b82572753afa91085fbf954f7c7613ff8376da9b26153d5cc3a3c4ee8R77
Set this to True to enable the caching and we should see significant speed up in the subsequent checkpoint save attempts, specially for larger scale jobs. Reference issue: https://github.com/pytorch/pytorch/issues/123695

## Experiment results:
```
Model size: 1.6 TB post dedupe, Ranks: 256
First checkpoint save time: 280s.
Subsequent checkpoint save time: 155s. 
E2e latency improvement: ~45%
```


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

kwen2501 commented:
> In next iteration, compare the local plan with the cached plan metadata. If no change, do not send that local plan in the collective.

Just curious, any chance different ranks have different decisions in this step?

saumishr commented:
> > In next iteration, compare the local plan with the cached plan metadata. If no change, do not send that local plan in the collective.
> 
> Just curious, any chance different ranks have different decisions in this step?

@kwen2501 Yes however the current logic takes care of that. SavePlans are built as per the state dict, the ranks need to save. Intuition here is that for a training job iteration, the state dict in multiple save attempts remains the same mostly and only the data changes. Therefore, cache hit rate will be very high, saving us a lot of cost on collectives. If the tensors change in any uncommon situation, ranks compare the new plan with the cached plan and send out the new plan over the collective.

meetv18 reviewed (commented):
@saumishr Wonder if we should add UTs with caching enabled?

meetv18 reviewed (changes requested):
@saumishr Wonder if we should add UTs with caching enabled?

meetv18 reviewed (approved):
Overall LGTM! Have some minor comments, would be great if we can resolve them before landing :)

saumishr committed:
[DCP] Cache save plans in default planner (#147343)

Summary:

This PR caches the save plans to significantly reduce the collective cost for successive checkpoint save attempts. Here is the high level approach:
-  Create the local plan and cache the same.
- In next iteration, compare the local plan with the cached plan metadata. If no change, do not send that local plan in the collective.
- Global plan step, will only create the global plan with the new delta plans and empty plans for the cached ones. 
- Finish plan step will check for the empty plans. If its empty, it will grab the cached plan. If not, it will use the new plan provided.

Test Plan: UTs

Reviewed By: MeetVadakkanchery

Differential Revision: D69224491

saumishr commented:
@pytorchbot  merge
